Forest

FridljDa opened the pull request:
Added forest functionality. I will revisit smart default hyper parameters at a later stage. Special focus on backwards compatibility: any past code written using IHW should be unaffected.
points of deviation in original code:

- ihw-class for results has more slots and most getters behave differently depending on which stratification method was used
- multidimensional covariates are also allowed for quantile slicing (e.g. per-covariate bins). Hence covariate is allowed as matrix input and ihw.formula has been adjusted

FridljDa committed:
ignore stuff

FridljDa committed:
introducing ihw_forest

FridljDa committed:
remove browser

FridljDa committed:
add nvar in auto nbins

FridljDa committed:
change nbins multidim

FridljDa committed:
change nbins multidim

FridljDa committed:
added no stratification method

FridljDa committed:
bins too small not allowed

const-ae commented on R/helpers.R:
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
+#'
+#'  Hypotheses are stratified into nbins different strata of (approximately) equal size based on
+#' increasing value of the covariate
+#' extension of `groups_by_filter` for multidimensional covariates, per-covariate quantiles
```
You are using `` ` `` to indicate code. This relies on markdown formatting which Roxygen does not used by default. You can turn it on by adding `Roxygen: list(markdown = TRUE)` to the `DESCRIPTION` file (see for example https://github.com/const-ae/glmGamPoi/blob/master/DESCRIPTION). However as Nikos seems to be rather using the explicit notation (i.e. `\code{\link{rank}}`), so I would recommend that you adopt the same style.

FridljDa replied:
Thank you, I changed it everywhere

const-ae commented on R/helpers.R:
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
+#' @param covariates Numeric vector of ordinal covariates based on which the stratification will be done.
+#' @param nbins Integer, number of groups/strata into which p-values will be split based on covariate.
+#' @param ties.method Character specifying how ties are treated, see \code{\link{rank}} function.
+#' @param seed Integer, specifies random seed to be used when ties.method=="random".
```
Wrap such code examples in `\code{ties.method=="random"}` so that they are rendered nicely.

FridljDa replied:
Thank you, I adressed this and also updated some documantation in `ihw_convex.R` accordingly

const-ae commented on R/helpers.R:
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
+#' increasing value of the covariate
+#' extension of `groups_by_filter` for multidimensional covariates, per-covariate quantiles
+#'
+#' @param covariates Numeric vector of ordinal covariates based on which the stratification will be done.
```
Is `covariates` really a numeric vector? Line 58 (`covariates[, i]`) suggests that it should be two dimensional?

FridljDa replied:
No it is indeed not. I updated the documantation.

const-ae commented on R/helpers.R:
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
+  
+  nvar <- ncol(covariates)
+  groups <- lapply(seq_len(nvar), function(i) {
+    covariate_i <- covariates[, i]
```
You have to be careful here. I assume you want your method to handle any kind of typical two-dimensional data structure (`matrix`, `data.frame`, and `tibble`). However tibbles behave a bit different: they don't automatically return a vector if you extract a single column:
``` r
tib <- tibble(a = letters[1:5], num = 1:5)
tib[,1]
#> # A tibble: 5 × 1
#>   a    
#>   <chr>
#> 1 a    
#> 2 b    
#> 3 c    
#> 4 d    
#> 5 e
tib[,1,drop=TRUE]
#> [1] "a" "b" "c" "d" "e"
```

FridljDa replied:
I specified `drop=TRUE` in this line

const-ae commented on R/helpers.R:
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
+    groups <- unname(unlist(groups))
+  }else{
+    groups <- apply(groups, 1, paste, collapse = "-") # base R equivalent of tidyr::unite
+  }
```
Why did you convert `groups` into a data.frame if you anyway reduce the rows back to a vector?

FridljDa replied:
Thank you for pointing this out, it was obsolete at this point!

const-ae commented on R/helpers.R:
```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
+  # convert to factor
+  groups <- as.factor(groups)
+  
+  return(groups)
```
You can leave out the explicit `return`

FridljDa replied:
done

const-ae commented on R/ihw_class.R:
```diff
@@ -47,13 +51,17 @@ ihwResult <- setClass("ihwResult",
+           		     solver_information= "list",
+           		     stratification_method = "character",
+           		     weight_matrices_forest = "list",
+           		     ntaus = "integer",
```
I would avoid having greek letter variable names in public-facing methods or objects. They only make sense if you know which formula the letter refers to. I would thus recommend renaming all occurrences of `ntaus` to some more literal variable name like `n_thresholds` or `n_censor_thres`

FridljDa replied:
I changed all `ntaus` to `n_censor_thres` in the whole project

const-ae commented on R/ihw_class.R:
```diff
@@ -142,7 +162,12 @@ setMethod("weighted_pvalues", signature(object="ihwResult"),
 
 covariates_ihwResult <- function(object){
-  object@df$covariate
+  covariate <- dplyr::select(object@df, dplyr::starts_with("covariate")) #TODO fix this
```
`dplyr` is only listed as a suggested dependency. You cannot use it in regular code without checking that it is indeed available.

const-ae replied:
You could replace the dplyr code by adapting the following snipped
``` r
dat <- data.frame(a_1 = 1:3, a_2 = 4:6, b_1 = letters[1:3], asdf = 0)
dat
#>   a_1 a_2 b_1 asdf
#> 1   1   4   a    0
#> 2   2   5   b    0
#> 3   3   6   c    0
sel_cols <- grepl("^a_", colnames(dat))
dat[,sel_cols]
#>   a_1 a_2
#> 1   1   4
#> 2   2   5
#> 3   3   6
```

const-ae commented on R/ihw_class.R:
```diff
@@ -167,8 +192,12 @@ setMethod("covariate_type", signature(object="ihwResult"),
-	object@df$group
+groups_factor_ihwResult <- function(object){#
+  if(object@stratification_method == "forest"){
+    groups <- dplyr::select(object@df, dplyr::contains("tree")) 
```
See above!

const-ae commented on R/plots.R:
```diff
@@ -27,90 +27,91 @@
          stop(sprintf("Invalid scale='%s'.", scale))
   ) 
-}  
+}  
```
Please try to fix the white space in this file so that it gets easier to review

FridljDa replied:
In the new commit, white space changes are finally not being displayed anymore (I was aware of this and had a few tries before)

const-ae reviewed (commented):
I went through parts of the code, and it looks quite solid to me. Of course, I cannot comment on the validity of the new random forest-based hypothesis weighting, but the code seems to be of high quality.

There are a few minor things that I have commented in-line; @nignatiadis feel free to correct me on anything that I have suggested :)


----------
Additional comments:

* What's up with the empty files?

FridljDa committed:
adressed most comments from pull request review

const-ae commented on R/ihw_class.R:
```diff
@@ -161,13 +161,10 @@ setMethod("weighted_pvalues", signature(object="ihwResult"),
-    as.matrix(covariate)
-  }
+covariates_ihwResult <- function(object){ 
+  covariate_names <- grepl("^covariate_", colnames(object@df))
```
Sorry that was probably a bad example. The underscore `_` had no particular meaning in that regex!

FridljDa committed:
plot white space fix (hopfully?)

FridljDa committed:
regex fix

FridljDa committed:
tested getters, they work

FridljDa committed:
comment out unit test

FridljDa committed:
more unit tests

FridljDa committed:
changed default ntrees

FridljDa committed:
add line to dealt with boundary case

FridljDa committed:
remove boundaries and unnecessary numerical instability

FridljDa committed:
updated documentation

FridljDa committed:
export stratification_breaks

FridljDa committed:
next try

FridljDa committed:
next try

FridljDa committed:
disbale autoset

FridljDa committed:
renamed nodedepth

FridljDa committed:
renamed stuff

FridljDa committed:
cleaning, smarter default parameter

FridljDa committed:
more documentation

FridljDa committed:
Merge remote-tracking branch 'origin/export_strat' into forest

FridljDa committed:
more diagnostic

FridljDa committed:
more diagnostic

FridljDa committed:
Merge branch 'export_strat' into forest

FridljDa committed:
cleaner documentation

FridljDa committed:
add gitignore

FridljDa committed:
nbin_dim

FridljDa committed:
uniform deviation instead of total variation

FridljDa committed:
allow smaller bins

FridljDa committed:
add nbins functionality to ihw forest

FridljDa committed:
ignore stuff

FridljDa committed:
added functionality to plotting

FridljDa committed:
removed white space

FridljDa committed:
comment plot function

Commit:
updated error handling

Commit:
Merge branch 'forest' of https://github.com/FridljDa/IHW into forest

FridljDa committed:
added handler

Commit:
default no regularisation for forest

Commit:
Merge branch 'forest' of https://github.com/FridljDa/IHW into forest

FridljDa committed:
default penalty for multidimensional covariates now uniform deviation

FridljDa committed:
a single threshold for tau

FridljDa committed:
single threshold in test

FridljDa committed:
removed all occurences of n_censor_thres

FridljDa committed:
remove n_censor

Commit:
white space changes

Commit:
remove n_censor_thres

Merge branch 'forest' of https://github.com/FridljDa/IHW into forest

# Conflicts:
#	R/ihw_forest.R

FridljDa committed:
new default parameters

FridljDa committed:
more documentation

FridljDa committed:
Merge branch 'forest' of https://github.com/FridljDa/IHW into forest

FridljDa committed:
new default

FridljDa committed:
new defaults
